include/alpaka/core/CallbackThread.hpp

```diff
@@ -34,46 +37,65 @@ namespace alpaka::core
                 m_thread.join();
             }
         }
-        auto submit(Task&& newTask) -> std::future<void>
+
+        // Note: due to different std lib implementations of packaged_task, the lifetime of the passed function either
+        // ends when the packaged_task is destroyed (while the returned future is still alive) or when both are
+        // destroyed.
+        template<typename NullaryFunction>
+        auto submit(NullaryFunction&& nf) -> std::future<void>
         {
-            auto f = newTask.get_future();
+            static_assert(
+                std::is_void_v<std::invoke_result_t<NullaryFunction>>,
+                "Submitted function must not have any arguments and return void.");
+            auto task = Task{std::forward<NullaryFunction>(nf)};
+            auto f = task.get_future();
             {
                 std::unique_lock<std::mutex> lock{m_mutex};
-                m_tasks.emplace(std::move(newTask));
+                ++m_tasksInProgress;
+                m_tasks.emplace(std::move(task));
                 if(!m_thread.joinable())
                     startWorkerThread();
             }
             m_cond.notify_one();
             return f;
         }
 
+        [[nodiscard]] auto empty() const
+        {
+            return m_tasksInProgress == 0;
+        }
+
     private:
         std::thread m_thread;
         std::condition_variable m_cond;
         std::mutex m_mutex;
-        std::atomic<bool> m_stop{false};
+        bool m_stop{false};
         std::queue<Task> m_tasks;
+        std::atomic<int> m_tasksInProgress{0};
 
         auto startWorkerThread() -> void
         {
             m_thread = std::thread(
                 [this]
                 {
-                    Task task;
```

Comment:
note: This is a bug I found yesterday. The problem is that the TAsk is always alive as long as the thread lives. This is moving the livetime of all objects stored in the task to the end of the call back thread or to the moment where the task is overwritten with a new tasks. The result of this is undefined and can create bugs.

Reply:
The changes were merged in #1976. I rebased on the top of the changes.